### PR TITLE
added bagelai.py

### PR DESCRIPTION
**Pull request by MoSBanapple**

### CODE REVIEW

File: bagelai.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import bagel
+import random
+
+class BagelAI:
```
Comment (qiwzhang) [inline bagelai.py]: It is better to use two separate classes for your two AI

File: bagelai.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    """
+    Guesses using the first AI
+    """
+    def Guess(self):
+        while None in self.digits:
+            num = bagel.Rand()
+            dig = bagel.Digitize(num)
+            for entry in dig:
+                if entry in self.digits:
+                    continue
```
Comment (qiwzhang) [inline bagelai.py]: not sure this "continue" is for line 19 or line 16.    please not use it.

Comment (qiwzhang) [inline bagelai.py]: Could you describe the algorithm for both AI?

File: bagelai.py [hunk 3]
```diff
@@ -6,29 +6,66 @@ class BagelAI:
...
+                return (digits[1], 1)
+            else:
+                return (digits[2], 0)
+                                                                                                                                                                                                                                    
+
 
+class BagelAI2:
+
+    def __init__(self):
+        self.number2 = bagel.Number(bagel.Rand())
```
Comment (qiwzhang) [inline bagelai.py]: why number2 ?     It can be just "number"

File: bagelai.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                self.digits[tup[1]] = tup[0]
+    """
+    Given a single pico or fermi, finds the correct digit and position and returns it in a tuple.
+    """
+    def FindFermi(self, num):
+        clues = self.number.Clue(num)
+        digits = bagel.Digitize(num)
+        if clues[0] != "Fermi":
+            digits = Swap(digits, 0, 1)
+            clues = self.number.Clue(Condense(digits))
```
Comment (qiwzhang) [inline bagelai.py]: no need Condense

can your Bagel class have two Clue functions, one  with array. and another one with number?

Comment (qiwzhang) [inline bagelai.py]: I feel line 35 to 44 can use a loop.

File: bagelai.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            return digits
+        digits = Swap(digits, 1, 2)
+        if self.number2.Clue(Condense(digits)) == ["Fermi","Fermi","Fermi"]:
+            return digits
+        digits = Swap(digits, 0, 2)
+        if self.number2.Clue(Condense(digits)) == ["Fermi","Fermi","Fermi"]:
+            return digits
+        digits = Swap(digits, 1, 2)
+        if self.number2.Clue(Condense(digits)) == ["Fermi","Fermi","Fermi"]:
+            return digits
```
Comment (qiwzhang) [inline bagelai.py]: line 75 yo 90 can use a loop,

File: bagelai.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            num = bagel.Rand()
+            dig = bagel.Digitize(num)
+            for entry in dig:
+                if entry in self.digits:
+                    repeat = True
+            if repeat:
+                repeat = False
+                continue
+            guess = self.number.Clue(num)
+            if len(guess) == 1 and guess[0] != "Bagel":
```
Comment (MoSBanapple) [inline bagelai.py]: The current guessing algorithm is designed to work with one Pico or Fermi, so it only takes those.

File: bagel.py [hunk 7]
```diff
@@ -29,6 +29,12 @@ def Clue(self, guess):
         else:
             clues.sort()
         return clues
+
+    """
+    Same as Clue but takes a digitized array instead
+    """
+    def ClueArray(self, guess):
```
Comment (qiwzhang) [inline bagel.py]: this is not efficient.   The number is combined here and digitized again in line 18.
I would make line 16 to take array, define a new function to take number

File: bagelai.py [hunk 8]
```diff
@@ -14,41 +14,45 @@ def __init__(self):
...
+                tup = self.FindFermi(num)
+                self.digits[tup[1]] = tup[0]
+
+    """
+    Returns a random number not containing digits already found.
+    """
+    def GetNonSeenRandom(self):
+        while True:
```
Comment (qiwzhang) [inline bagelai.py]: you can do:

valid = False
while !valid:
...
         
       if entry in self.digits:
             valid = false
             break

File: bagelai.py [hunk 9]
```diff
@@ -14,41 +14,45 @@ def __init__(self):
...
-            digits = Swap(digits, 0, 2)
-            digits = Swap(digits, 1, 2)
-        digits = Swap(digits, 0, 1)
-        if self.number.Clue(Condense(digits)) == ["Fermi"]:
+        temp = digits
+        for entry in [(0,1),(0,2),(1,2)]:
+            if clues[0] == "Fermi":
+                break
+            temp = Swap(digits, entry[0], entry[1])
+            clues = self.number.ClueArray(digits)
```
Comment (qiwzhang) [inline bagelai.py]: should this line be:

clues = self.number.ClueArray(temp)

otherwise temp is not used

**Pull request merged**